File: backend/src/routes/files.ts
import { Hono } from 'hono';
import { promises as fs } from 'fs';
import path from 'path';
import { applySearchReplace, changedLineCount } from '../utils/edits';
import { isProtectedPath, isProtectedDirName } from '../utils/protected-dirs';
import { isPathInside } from '../utils/containment';
import type { Variables } from '../types';

const files = new Hono<{ Variables: Variables }>();

// ─── Workspace sandbox ──────────────────────────────────────────
// The files API is reachable by any authenticated client over the network,
// so every content/delete/write request must name a workspace root and the
// target path must resolve INSIDE it. Directory listing stays open (the
// workspace picker needs to browse), but reading/writing/deleting is
// confined to the workspace the client declares.
function resolveWorkspaceRoot(ws?: string): string | null {
  if (!ws || typeof ws !== 'string') return null;
  const resolved = path.resolve(ws);
  // Reject drive roots (C:\ or /) — a workspace must be a real subfolder,
  // otherwise "workspacePath=C:\" would unlock the whole disk.
  if (path.parse(resolved).root === resolved) return null;
  return resolved;
}

const IGNORE_DIRS = new Set([
  'node_modules', '.git', '.svn', '.hg', '.DS_Store',
  '__pycache__', '.next', '.nuxt', 'dist', 'build', '.cache',
  'target', 'vendor', '.venv', 'venv', 'env',
]);

const MAX_FILE_SIZE = 1024 * 1024; // 1MB max for preview

// Detect language from file extension
function detectLanguage(filename: string): string | null {
  const ext = path.extname(filename).toLowerCase();
  const map: Record<string, string> = {
    '.ts': 'typescript', '.tsx': 'tsx', '.js': 'javascript', '.jsx': 'jsx',
    '.json': 'json', '.md': 'markdown', '.css': 'css', '.scss': 'scss',
    '.html': 'html', '.xml': 'xml', '.svg': 'xml', '.yaml': 'yaml', '.yml': 'yaml',
    '.py': 'python', '.rb': 'ruby', '.go': 'go', '.rs': 'rust', '.c': 'c', '.cpp': 'cpp',
    '.h': 'c', '.hpp': 'cpp', '.java': 'java', '.kt': 'kotlin', '.swift': 'swift',
    '.sh': 'bash', '.bash': 'bash', '.zsh': 'bash', '.ps1': 'powershell',
    '.sql': 'sql', '.graphql': 'graphql', '.gql': 'graphql',
    '.dockerfile': 'dockerfile', '.txt': 'text', '.env': 'text',
    '.toml': 'ini', '.ini': 'ini', '.cfg': 'ini',
    '.vue': 'vue', '.svelte': 'html', '.astro': 'html',
  };
  return map[ext] || null;
}

// Check if content appears to be binary
function isBinaryContent(buffer: Buffer): boolean {
  const sampleSize = Math.min(buffer.length, 8192);
  for (let i = 0; i < sampleSize; i++) {
    if (buffer[i] === 0) return true;
  }
  return false;
}

files.get('/', async (c) => {
  const dirPath = c.req.query('path');
  if (!dirPath) {
    return c.json({ error: 'path query parameter is required' }, 400);
  }

  // Security: resolve the path and ensure it's absolute
  const resolved = path.resolve(dirPath);

  try {
    await fs.access(resolved);
  } catch {
    return c.json({ error: 'Directory does not exist' }, 404);
  }

  const stat = await fs.stat(resolved);
  if (!stat.isDirectory()) {
    return c.json({ error: 'Path is not a directory' }, 400);
  }

  try {
    const entries = await fs.readdir(resolved, { withFileTypes: true });

    const result = await Promise.all(
      entries
        .filter((entry) => !entry.name.startsWith('.') && !IGNORE_DIRS.has(entry.name) && !isProtectedDirName(entry.name))
        .map(async (entry) => {
          const fullPath = path.join(resolved, entry.name);
          let size: number | undefined;
          if (entry.isFile()) {
            try {
              const fileStat = await fs.stat(fullPath);
              size = fileStat.size;
            } catch { /* skip */ }
          }
          return {
            name: entry.name,
            path: fullPath,
            type: entry.isDirectory() ? 'directory' : 'file',
            size,
          };
        })
    );

    // Sort: directories first, then files, alphabetically
    result.sort((a, b) => {
      if (a.type !== b.type) return a.type === 'directory' ? -1 : 1;
      return a.name.localeCompare(b.name);
    });

    return c.json({ entries: result });
  } catch (e) {
    return c.json({ error: 'Failed to read directory' }, 500);
  }
});

files.get('/content', async (c) => {
  const filePath = c.req.query('path');
  if (!filePath) {
    return c.json({ error: 'path query parameter is required' }, 400);
  }

  const workspaceRoot = resolveWorkspaceRoot(c.req.query('workspacePath'));
  if (!workspaceRoot) {
    return c.json({ error: 'A valid workspacePath query parameter is required' }, 403);
  }

  const resolved = path.resolve(filePath);
  if (isProtectedPath(workspaceRoot, resolved)) {
    return c.json({ error: 'Access denied: path is in a protected server directory' }, 403);
  }
  if (!(await isPathInside(workspaceRoot, resolved))) {
    return c.json({ error: 'Access denied: path is outside the workspace' }, 403);
  }

  try {
    await fs.access(resolved);
  } catch {
    return c.json({ error: 'File does not exist' }, 404);
  }

  const stat = await fs.stat(resolved);
  if (!stat.isFile()) {
    return c.json({ error: 'Path is not a file' }, 400);
  }

  const truncated = stat.size > MAX_FILE_SIZE;

  try {
    const buffer = await fs.readFile(resolved, { 
      flag: 'r',
    });

    if (isBinaryContent(buffer)) {
      return c.json({
        content: null,
        language: null,
        size: stat.size,
        truncated: false,
        binary: true,
      });
    }

    const content = truncated
      ? buffer.subarray(0, MAX_FILE_SIZE).toString('utf-8')
      : buffer.toString('utf-8');

    return c.json({
      content,
      language: detectLanguage(filePath),
      size: stat.size,
      truncated,
      binary: false,
    });
  } catch (e) {
    return c.json({ error: 'Failed to read file' }, 500);
  }
});

files.delete('/delete', async (c) => {
  const filePath = c.req.query('path');
  if (!filePath) {
    return c.json({ error: 'path query parameter is required' }, 400);
  }

  const workspaceRoot = resolveWorkspaceRoot(c.req.query('workspacePath'));
  if (!workspaceRoot) {
    return c.json({ error: 'A valid workspacePath query parameter is required' }, 403);
  }

  const resolved = path.resolve(filePath);
  if (isProtectedPath(workspaceRoot, resolved)) {
    return c.json({ error: 'Access denied: path is in a protected server directory' }, 403);
  }
  if (!(await isPathInside(workspaceRoot, resolved))) {
    return c.json({ error: 'Access denied: path is outside the workspace' }, 403);
  }

  try {
    await fs.access(resolved);
    const stat = await fs.stat(resolved);
    if (stat.isDirectory()) {
      await fs.rm(resolved, { recursive: true });
    } else {
      await fs.unlink(resolved);
    }
    return c.json({ success: true, path: resolved });
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return c.json({ error: 'File does not exist' }, 404);
    }
    return c.json({ error: 'Failed to delete file' }, 500);
  }
});

files.post('/edit', async (c) => {
  try {
    const { filePath, oldString, newString, workspacePath } = await c.req.json();
    if (!filePath || typeof oldString !== 'string' || typeof newString !== 'string') {
      return c.json({ error: 'filePath, oldString and newString are required' }, 400);
    }

    const workspaceRoot = resolveWorkspaceRoot(workspacePath);
    if (!workspaceRoot) {
      return c.json({ error: 'A valid workspacePath is required in the request body' }, 403);
    }

    const resolved = path.resolve(filePath);
    if (isProtectedPath(workspaceRoot, resolved)) {
      return c.json({ error: 'Access denied: path is in a protected server directory' }, 403);
    }
    if (!(await isPathInside(workspaceRoot, resolved))) {
      return c.json({ error: 'Access denied: path is outside the workspace' }, 403);
    }

    let oldContent: string;
    try {
      oldContent = await fs.readFile(resolved, 'utf-8');
    } catch {
      return c.json({ error: 'File does not exist' }, 404);
    }

    const result = applySearchReplace(oldContent, oldString, newString);
    if (!result.ok || result.newContent === undefined) {
      return c.json({ error: result.error || 'Edit failed' }, 400);
    }

    await fs.writeFile(resolved, result.newContent, 'utf-8');

    return c.json({
      success: true,
      path: resolved,
      size: Buffer.byteLength(result.newContent, 'utf-8'),
    });
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : 'Failed to edit file' }, 500);
  }
});

files.put('/write', async (c) => {
  try {
    const { filePath, content, workspacePath } = await c.req.json();
    if (!filePath || content === undefined) {
      return c.json({ error: 'filePath and content are required' }, 400);
    }

    const workspaceRoot = resolveWorkspaceRoot(workspacePath);
    if (!workspaceRoot) {
      return c.json({ error: 'A valid workspacePath is required in the request body' }, 403);
    }

    const resolved = path.resolve(filePath);
    if (isProtectedPath(workspaceRoot, resolved)) {
      return c.json({ error: 'Access denied: path is in a protected server directory' }, 403);
    }
    if (!(await isPathInside(workspaceRoot, resolved))) {
      return c.json({ error: 'Access denied: path is outside the workspace' }, 403);
    }

    // Ensure parent directory exists
    const parentDir = path.dirname(resolved);
    await fs.mkdir(parentDir, { recursive: true });

    // Read old content for diff (if file exists)
    let oldContent: string | null = null;
    try {
      const oldBuffer = await fs.readFile(resolved);
      oldContent = oldBuffer.toString('utf-8');
    } catch { /* file doesn't exist yet */ }

    if (oldContent !== null) {
      // EXISTING FILE: apply a SURGICAL diff so the write only changes the
      // lines that actually differ (same rule as the agent's write_file tool).
      // A version that rewrites most of the file is refused — otherwise a
      // slightly-off full-file re-emit would silently clobber the user's work.
      const { count: changed, total } = changedLineCount(
        oldContent.replace(/\r\n/g, '\n'),
        content.replace(/\r\n/g, '\n')
      );
      const isSmallEdit = changed <= Math.max(20, Math.floor(total * 0.4));
      if (!isSmallEdit) {
        return c.json({ error: `Refusing to overwrite ${filePath}: your version changes ${changed} of ${total} lines — that is a full rewrite, not an edit. To rewrite the whole file on purpose, delete it first (or use the edit/EDIT flow with a small old_string).` }, 409);
      }
      // Preserve the file's existing line-ending style.
      const finalContent = oldContent.includes('\r\n')
        ? content.replace(/\r\n/g, '\n').replace(/\n/g, '\r\n')
        : content;
      await fs.writeFile(resolved, finalContent, 'utf-8');
      return c.json({
        success: true,
        path: resolved,
        isNew: false,
        size: Buffer.byteLength(finalContent, 'utf-8'),
      });
    }

    await fs.writeFile(resolved, content, 'utf-8');

    return c.json({
      success: true,
      path: resolved,
      isNew: true,
      size: Buffer.byteLength(content, 'utf-8'),
    });
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : 'Failed to write file' }, 500);
  }
});

export default files;

File: backend/app/routes/files.py
"""Files API — mirrors backend/src/routes/files.ts.

Workspace-sandboxed file listing, reading, editing, writing, deleting.
Directory listing stays open (the workspace picker needs to browse), but
reading/writing/deleting is confined to the declared workspace root.
"""

from __future__ import annotations

import os
from pathlib import Path

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..utils import apply_search_replace, changed_line_count, is_path_inside, is_protected_dir_name, is_protected_path

router = APIRouter()

IGNORE_DIRS = {
    "node_modules", ".git", ".svn", ".hg", ".DS_Store",
    "__pycache__", ".next", ".nuxt", "dist", "build", ".cache",
    "target", "vendor", ".venv", "venv", "env",
}

MAX_FILE_SIZE = 1024 * 1024  # 1MB max for preview

LANGUAGE_MAP = {
    ".ts": "typescript", ".tsx": "tsx", ".js": "javascript", ".jsx": "jsx",
    ".json": "json", ".md": "markdown", ".css": "css", ".scss": "scss",
    ".html": "html", ".xml": "xml", ".svg": "xml", ".yaml": "yaml", ".yml": "yaml",
    ".py": "python", ".rb": "ruby", ".go": "go", ".rs": "rust", ".c": "c", ".cpp": "cpp",
    ".h": "c", ".hpp": "cpp", ".java": "java", ".kt": "kotlin", ".swift": "swift",
    ".sh": "bash", ".bash": "bash", ".zsh": "bash", ".ps1": "powershell",
    ".sql": "sql", ".graphql": "graphql", ".gql": "graphql",
    ".dockerfile": "dockerfile", ".txt": "text", ".env": "text",
    ".toml": "ini", ".ini": "ini", ".cfg": "ini",
    ".vue": "vue", ".svelte": "html", ".astro": "html",
}


def _resolve_workspace_root(ws: str | None) -> str | None:
    if not ws or not isinstance(ws, str):
        return None
    resolved = os.path.abspath(ws)
    if os.path.splitdrive(resolved)[1] in ("\\", "/") or resolved == "/":
        return None  # reject drive roots
    return resolved


def _detect_language(filename: str) -> str | None:
    ext = os.path.splitext(filename)[1].lower()
    return LANGUAGE_MAP.get(ext)


def _is_binary_content(data: bytes) -> bool:
    sample = data[:8192]
    return b"\x00" in sample


@router.get("")
async def list_dir(request: Request) -> dict:
    dir_path = request.query_params.get("path")
    if not dir_path:
        return JSONResponse({"error": "path query parameter is required"}, status_code=400)
    resolved = os.path.abspath(dir_path)
    if not os.path.exists(resolved):
        return JSONResponse({"error": "Directory does not exist"}, status_code=404)
    if not os.path.isdir(resolved):
        return JSONResponse({"error": "Path is not a directory"}, status_code=400)
    try:
        entries = os.listdir(resolved)
        result = []
        for name in entries:
            if name.startswith(".") or name in IGNORE_DIRS or is_protected_dir_name(name):
                continue
            full = os.path.join(resolved, name)
            size = None
            if os.path.isfile(full):
                try:
                    size = os.path.getsize(full)
                except OSError:  # noqa: S110
                    pass
            result.append(
                {
                    "name": name,
                    "path": full,
                    "type": "directory" if os.path.isdir(full) else "file",
                    "size": size,
                }
            )
        result.sort(key=lambda e: (e["type"] != "directory", e["name"].lower()))
        return {"entries": result}
    except OSError:
        return JSONResponse({"error": "Failed to read directory"}, status_code=500)


@router.get("/content")
async def get_content(request: Request) -> dict:
    file_path = request.query_params.get("path")
    if not file_path:
        return JSONResponse({"error": "path query parameter is required"}, status_code=400)
    workspace_root = _resolve_workspace_root(request.query_params.get("workspacePath"))
    if not workspace_root:
        return JSONResponse({"error": "A valid workspacePath query parameter is required"}, status_code=403)
    resolved = os.path.abspath(file_path)
    if is_protected_path(workspace_root, resolved):
        return JSONResponse({"error": "Access denied: path is in a protected server directory"}, status_code=403)
    if not (await is_path_inside(workspace_root, resolved)):
        return JSONResponse({"error": "Access denied: path is outside the workspace"}, status_code=403)
    if not os.path.exists(resolved):
        return JSONResponse({"error": "File does not exist"}, status_code=404)
    if not os.path.isfile(resolved):
        return JSONResponse({"error": "Path is not a file"}, status_code=400)
    stat = os.stat(resolved)
    truncated = stat.st_size > MAX_FILE_SIZE
    try:
        with open(resolved, "rb") as f:
            buffer = f.read()
        if _is_binary_content(buffer):
            return {"content": None, "language": None, "size": stat.st_size, "truncated": False, "binary": True}
        content = buffer[:MAX_FILE_SIZE].decode("utf-8", "replace") if truncated else buffer.decode("utf-8", "replace")
        return {"content": content, "language": _detect_language(file_path), "size": stat.st_size, "truncated": truncated, "binary": False}
    except OSError:
        return JSONResponse({"error": "Failed to read file"}, status_code=500)


@router.delete("/delete")
async def delete_file_route(request: Request) -> dict:
    file_path = request.query_params.get("path")
    if not file_path:
        return JSONResponse({"error": "path query parameter is required"}, status_code=400)
    workspace_root = _resolve_workspace_root(request.query_params.get("workspacePath"))
    if not workspace_root:
        return JSONResponse({"error": "A valid workspacePath query parameter is required"}, status_code=403)
    resolved = os.path.abspath(file_path)
    if is_protected_path(workspace_root, resolved):
        return JSONResponse({"error": "Access denied: path is in a protected server directory"}, status_code=403)
    if not (await is_path_inside(workspace_root, resolved)):
        return JSONResponse({"error": "Access denied: path is outside the workspace"}, status_code=403)
    try:
        if not os.path.exists(resolved):
            return JSONResponse({"error": "File does not exist"}, status_code=404)
        if os.path.isdir(resolved):
            import shutil

            shutil.rmtree(resolved)
        else:
            os.remove(resolved)
        return {"success": True, "path": resolved}
    except OSError:
        return JSONResponse({"error": "Failed to delete file"}, status_code=500)


@router.post("/edit")
async def edit_file_route(request: Request) -> dict:
    try:
        body = await request.json()
        file_path = body.get("filePath")
        old_string = body.get("oldString")
        new_string = body.get("newString")
        if not file_path or not isinstance(old_string, str) or not isinstance(new_string, str):
            return JSONResponse({"error": "filePath, oldString and newString are required"}, status_code=400)
        workspace_root = _resolve_workspace_root(body.get("workspacePath"))
        if not workspace_root:
            return JSONResponse({"error": "A valid workspacePath is required in the request body"}, status_code=403)
        resolved = os.path.abspath(file_path)
        if is_protected_path(workspace_root, resolved):
            return JSONResponse({"error": "Access denied: path is in a protected server directory"}, status_code=403)
        if not (await is_path_inside(workspace_root, resolved)):
            return JSONResponse({"error": "Access denied: path is outside the workspace"}, status_code=403)
        try:
            with open(resolved, "r", encoding="utf-8") as f:
                old_content = f.read()
        except OSError:
            return JSONResponse({"error": "File does not exist"}, status_code=404)
        result = apply_search_replace(old_content, old_string, new_string)
        if not result.get("ok") or result.get("newContent") is None:
            return JSONResponse({"error": result.get("error") or "Edit failed"}, status_code=400)
        with open(resolved, "w", encoding="utf-8") as f:
            f.write(result["newContent"])
        return {"success": True, "path": resolved, "size": len(result["newContent"].encode("utf-8"))}
    except Exception as e:  # noqa: BLE001
        return JSONResponse({"error": str(e) or "Failed to edit file"}, status_code=500)


@router.put("/write")
async def write_file_route(request: Request) -> dict:
    try:
        body = await request.json()
        file_path = body.get("filePath")
        content = body.get("content")
        if not file_path or content is None:
            return JSONResponse({"error": "filePath and content are required"}, status_code=400)
        workspace_root = _resolve_workspace_root(body.get("workspacePath"))
        if not workspace_root:
            return JSONResponse({"error": "A valid workspacePath is required in the request body"}, status_code=403)
        resolved = os.path.abspath(file_path)
        if is_protected_path(workspace_root, resolved):
            return JSONResponse({"error": "Access denied: path is in a protected server directory"}, status_code=403)
        if not (await is_path_inside(workspace_root, resolved)):
            return JSONResponse({"error": "Access denied: path is outside the workspace"}, status_code=403)
        os.makedirs(os.path.dirname(resolved), exist_ok=True)
        old_content: str | None = None
        try:
            with open(resolved, "r", encoding="utf-8") as f:
                old_content = f.read()
        except OSError:  # noqa: S110
            pass
        if old_content is not None:
            cc = changed_line_count(old_content.replace("\r\n", "\n"), content.replace("\r\n", "\n"))
            changed = cc["count"]
            total = cc["total"]
            is_small_edit = changed <= max(20, int(total * 0.4))
            if not is_small_edit:
                return JSONResponse(
                    {"error": f"Refusing to overwrite {file_path}: your version changes {changed} of {total} lines — that is a full rewrite, not an edit. To rewrite the whole file on purpose, delete it first (or use the edit/EDIT flow with a small old_string)."},
                    status_code=409,
                )
            final_content = content.replace("\r\n", "\n").replace("\n", "\r\n") if "\r\n" in old_content else content
            with open(resolved, "w", encoding="utf-8") as f:
                f.write(final_content)
            return {"success": True, "path": resolved, "isNew": False, "size": len(final_content.encode("utf-8"))}
        with open(resolved, "w", encoding="utf-8") as f:
            f.write(content)
        return {"success": True, "path": resolved, "isNew": True, "size": len(content.encode("utf-8"))}
    except Exception as e:  # noqa: BLE001
        return JSONResponse({"error": str(e) or "Failed to write file"}, status_code=500)